tests/test_web_poet_rules.py

```diff
@@ -0,0 +1,231 @@
+"""This tests scrapy-poet's integration with web-poet most especially when
+populating override settings via:
+
+..code-block::
+    from web_poet import default_registry
+
+    SCRAPY_POET_OVERRIDES = default_registry.get_rules()
+"""
+
+from typing import Any
+
+from pytest_twisted import inlineCallbacks
+from url_matcher import Patterns
+from web_poet import ApplyRule, default_registry
+
+from tests.po_lib import (
+    PORT,
+    URL,
+    ParentProduct,
+    ParentProductPage,
+    ParentReplacedProduct,
+    ParentReplacedProductPage,
+    POIntegrationPage,
+    POOverridenPage,
+    Product,
+    ProductFromInjectable,
+    ProductFromInjectablePage,
+    ProductPage,
+    ReplacedProduct,
+    ReplacedProductPage,
+    StandaloneProduct,
+    StandaloneProductPage,
+    SubclassProductPage,
+    SubclassReplacedProduct,
+    SubclassReplacedProductPage,
+)
+from tests.test_middleware import ProductHtml, spider_for
+from tests.utils import crawl_single_item, create_scrapy_settings
+
+# The rules are defined in `tests/po_lib/__init__.py`.
+RULES = default_registry.get_rules()
+
+
+def rules_settings() -> dict:
+    settings = create_scrapy_settings(None)
+    # Converting it to a set removes potential duplicate ApplyRules
+    settings["SCRAPY_POET_OVERRIDES"] = set(RULES)
+    return settings
+
+
+@inlineCallbacks
+def crawl_item(PageObject) -> Any:
+    """Helper function to easily return the item from a simulated Scrapy callback
+    which asks for either of these dependencies:
+        - Page Object
+        - Item Type
+    """
+    item, _, _ = yield crawl_single_item(
+        spider_for(PageObject), ProductHtml, rules_settings(), port=PORT
+    )
+    return item
+
+
+@inlineCallbacks
+def test_overrides() -> None:
+    """This tests the rules where a PageObject is used in lieu of another."""
+    item = yield crawl_item(POOverridenPage)
+    assert item == {"msg": "PO replacement"}
+
+
+@inlineCallbacks
+def test_item_return() -> None:
+    """This tests the rules where an Item Type is requested.
+
+    A corresponding ``ApplyRule`` should be available to denote that a Page
+    Object is equipped to produce it for the given URL pattern and Item Type.
+
+    For example, the code below should return a ``Product`` item which was taken
+    using the route ``Returns[ItemType]``.
+
+    ..code-block::
+
+        @handle_urls(URL)
+        class ProductPage(ItemPage[Product]):
+            ...
+    """
+    item = yield crawl_item(Product)
+    assert item == Product(name="product's name")
+
+
+@inlineCallbacks
+def test_item_return_subclass() -> None:
+    """A Page Object should properly inherit the ``Return[ItemType]`` from its
+    parent.
+
+    For the example below, asking for ParentProduct would return ``ParentProduct``
+    from the ``ParentProductPage`` since the rules have equal priority but the
+    ``ParentProductPage`` has been decleared first.
```

Comment:
I think we might need to issue a warning in this case, suggesting to set priority explicitly. What do you think?

Reply:
That's a good idea, as it could remove some confusion in users of why their rules aren't working as expected.

I've been writing tests on this but encountered the problem wherein the tests are flaky since the `ApplyRules` are being loaded in random orders for each run, affecting the expected order of priority should all of them have the default `priority=500`.

I think this could also be a problem for users later on, since the warning message changes each time.

I'll solve this problem first in the registry to make the order of rules deterministic.